presto-main/src/main/java/io/prestosql/operator/scalar/DateTimeFunctions.java

```diff
@@ -148,25 +148,37 @@ public static long fromUnixTime(@SqlType(StandardTypes.DOUBLE) double unixTime,
     }
 
     @ScalarFunction("from_unixtime_nanos")
-    @SqlType("timestamp(9)")
-    public static LongTimestamp fromUnixTimeNanos(@SqlType(StandardTypes.BIGINT) long unixTimeNanos)
-    {
-        long epochSeconds = unixTimeNanos / NANOSECONDS_PER_SECOND;
-        long nanosOfSecond = unixTimeNanos % NANOSECONDS_PER_SECOND;
-        long picosOfSecond = nanosOfSecond * PICOSECONDS_PER_NANOSECOND;
-        return DateTimes.longTimestamp(epochSeconds, picosOfSecond);
+    public static final class FromUnixtimeNanosDecimal
+    {
+        private FromUnixtimeNanosDecimal() {}
+
+        @LiteralParameters({"p", "s"})
+        @SqlType("timestamp(9)")
+        public static LongTimestamp fromLong(@LiteralParameter("s") long scale, @SqlType("decimal(p, s)") Slice unixTimeNanos)
+        {
+            BigInteger unixTimeNanosInt = new BigDecimal(Decimals.toString(unixTimeNanos, (int) scale)).toBigInteger();
+            long epochSeconds = unixTimeNanosInt.divide(BigInteger.valueOf(NANOSECONDS_PER_SECOND)).longValue();
+            long nanosOfSecond = unixTimeNanosInt.mod(BigInteger.valueOf(NANOSECONDS_PER_SECOND)).longValue();
+            long picosOfSecond = nanosOfSecond * PICOSECONDS_PER_NANOSECOND;
+            return DateTimes.longTimestamp(epochSeconds, picosOfSecond);
+        }
+
+        @LiteralParameters({"p", "s"})
+        @SqlType("timestamp(9)")
+        public static LongTimestamp fromShort(@SqlType("decimal(p, s)") long unixTimeNanos)
+        {
+            long epochSeconds = unixTimeNanos / NANOSECONDS_PER_SECOND;
```

Comment:
This code would only work for `s==0`. If there is decimal dot in the number we need to take that into account and basically round the value to nearest integer before proceeding with conversion to timestamp.
You can replicate rounding code from https://github.com/prestosql/presto/blob/2379c52ead0ca9f50d89fd75aa25e885a81118fa/presto-main/src/main/java/io/prestosql/operator/scalar/MathFunctions.java#L853-L870